eZ/Publish/Core/Repository/SearchService.php

```diff
@@ -344,8 +346,18 @@ protected function addPermissionsCriterion(Criterion &$criterion)
 
     public function supports(int $capabilityFlag): bool
     {
-        if ($this->searchHandler instanceof Capable) {
-            return $this->searchHandler->supports($capabilityFlag);
+        $handler = $this->searchHandler;
+
+        if ($handler instanceof ValueHolderInterface) {
+            if ($handler instanceof LazyLoadingInterface && !$handler->isProxyInitialized()) {
+                $handler->initializeProxy();
+            }
+
+            $handler = $handler->getWrappedValueHolderValue();
+        }
+
+        if ($handler instanceof Capable) {
+            return $handler->supports($capabilityFlag);
         }
```

Comment:
This is gonna be an issue for a lot of cases due to BC policies we have. I was rather thinking to delegate this decision to compiler pass magic in this instance.
We need an interface documented to be non-bc-safe which extends both Handler and Capable.

So, back to the interface - compiler pass should use it to instruct lazy proxy how to define service, if that service is indeed capable, it should substitute it. Not sure if it's gonna work the way I want though.

_Capable should've been part of Handler since 3.0, but I didn't manage to find time to merge interfaces (I didn't find time to do a lot of things)._

Reply:
> This is gonna be an issue for a lot of cases due to BC policies we have. I was rather thinking to delegate this decision to compiler pass magic in this instance.

I don't think it's possible. 

> We need an interface documented to be non-bc-safe which extends both Handler and Capable.

If you have in mind changing `ezpublish.spi.search_engine` definition to 

```
    ezpublish.spi.search_engine:
        class: eZ\Publish\SPI\Search\Engine
        factory: ["@ezpublish.api.search_engine.factory", buildSearchEngine]
        public: false
        lazy: true
```

where `eZ\Publish\SPI\Search\Engine` is

```
interface Engine extends Handler, Capable
{
}
```

it will not gonna work. If `\eZ\Publish\SPI\Search\Handler` implementation doesn't implement `\eZ\Publish\SPI\Search\Capable` (e.g. LSE) then calling \``\eZ\Publish\SPI\Search\Capable::supports` will throw **UndefinedMethodError**

```
Attempted to call an undefined method named "supports" of class "ContainerQXIx3RC\Handler_b009400".
```

> _Capable should've been part of Handler since 3.0, but I didn't manage to find time to merge interfaces (I didn't find time to do a lot of things)._

Agree.